src/Mvc/Mvc.RazorPages/src/Infrastructure/DefaultPageLoader.cs

```diff
@@ -41,16 +47,31 @@ internal class DefaultPageLoader : IPageLoader
 
         private IViewCompiler Compiler => _viewCompilerProvider.GetCompiler();
 
-        public CompiledPageActionDescriptor Load(PageActionDescriptor actionDescriptor)
+        public async override ValueTask<CompiledPageActionDescriptor> LoadAsync(PageActionDescriptor actionDescriptor)
         {
             if (actionDescriptor == null)
             {
                 throw new ArgumentNullException(nameof(actionDescriptor));
             }
 
-            var compileTask = Compiler.CompileAsync(actionDescriptor.RelativePath);
-            var viewDescriptor = compileTask.GetAwaiter().GetResult();
+            if (_memoryCache.TryGetValue(actionDescriptor, out CompiledPageActionDescriptor compiledDescriptor))
```

Comment:
This isn't atomic - does that matter?

Reply:
Who was caching this before? This is wierd that we're just adding another random cache, and it's double-wierd that it's using the change tokens as well.

You have cache in the compiler, cache here, and cache in the invoker provider. It seems like this is getting out of hand.

Reply:
This design made sense before because the cache in the compiler caches each type, and the cache in the invoker provide caches each AD. 

Should we just obsolete the loader entirely and make the invoker provider be the cache?